File: assign2/Lua/src/Factorial.lua
local fun = require 'src.fun'

function iterativeFactorial(number)
  if (number < 2) then
    return 1
  else
    return fun.product(fun.range(1, number))
  end
end

function recursiveFactorial(number)
  if (number < 2) then
	  return 1
  else
	  return number * recursiveFactorial(number - 1)
  end
end

function tailRecursiveFactorial(number, currentResult)
  if (number < 2) then
	  return (currentResult or 1)
  else
	  return tailRecursiveFactorial(number - 1, number * (currentResult or 1))
  end
end

File: assign2/Haskell/Src/Factorial.hs
module Src.Factorial where

factorialIterative givenNumber = product [1..givenNumber]

regularRecursiveFactorial 0 = 1
regularRecursiveFactorial givenNumber = givenNumber * regularRecursiveFactorial (givenNumber - 1)

tailRecursiveFactorial givenNumber = tailRecursive givenNumber 1 where
  tailRecursive givenNumber factorial = if givenNumber <= 1 then factorial else tailRecursive (givenNumber - 1) (factorial * givenNumber)
